flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -994,10 +972,18 @@ TimerService getTimerService() {
 		return timerService;
 	}
 
-	public ProcessingTimeService getProcessingTimeService(int operatorIndex) {
+	@VisibleForTesting
+	StreamOperator<?> getHeadOperator() {
+		return operatorChain.getHeadOperator();
+	}
+
+	public ProcessingTimeService getProcessingTimeService(StreamOperator<?> operator, int chainIndex) {
```

Comment:
We can find operator by its index in chain, no need to add an additional parameter.

Reply:
`chainIndex` here is obtained through `Streamconfig#getChainIndex` (which is generated when compiling into the job graph). I don't think it can be guaranteed that `chainIndex` is consistent with the index in the `OperatorChain#alloperators` array. At present, there is no data structure mapping `chainIndex` to `StreamOperator` in `OperatorChain`, while adding this structure to the `OperatorChain` seems to complicate things. I didn't think of a better way. Do you have other suggestions?

Reply:
Using operatorIndex as key in processingTimeServices map should work.

Reply:
The parameter passed in `getprocessingTimeService` is `chainindex`, but also there is no data structure mapping `chainIndex` to `operatorIndex ` in `OperatorChain`.

Reply:
Seems you are right, we need to pass 2 parameters.